### PR TITLE
Refs [TASK][36303] :Remote Data Source

### PR DESCRIPTION
**Pull request by hideoncalm**
## Related Tickets
- [#Task 36305: Create request to fetch data](https://edu-redmine.sun-asterisk.vn/issues/36305)
- [#Task 36306: Create remote data source](https://edu-redmine.sun-asterisk.vn/issues/36305)
## WHAT
- 
## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://github.com/framgia/coding-standards/tree/master/eng/android |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/coinqapp/data/source/remote/CoinRemoteDataSource.kt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import com.sunasterisk.coinqapp.utils.parseJsonToObject
+import org.json.JSONArray
+
+class CoinRemoteDataSource private constructor() : CoinDataSource.Remote {
+
+    override fun getCoins(
+        vs: String,
+        order: String,
+        pp: Int,
+        p: Int,
```
Comment (thonc-2135) [inline app/src/main/java/com/sunasterisk/coinqapp/data/source/remote/CoinRemoteDataSource.kt]: vs là j, pp , p là gì viết rõ hơn chút đk k

File: app/src/main/java/com/sunasterisk/coinqapp/data/source/remote/CoinRemoteDataSource.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    private fun getCoins(vs: String, order: String, pp: Int, p: Int, sparkline: Boolean, changeDay: String): List<Coin>{
+        val jsonString = getJsonFromUrl(APIQuery.queryCoins(vs, order, pp, p, sparkline, changeDay))
+        val jsonArray = JSONArray(jsonString)
+        return jsonArray.parseJsonToObject()
+    }
+
+    companion object {
+        private var instance: CoinRemoteDataSource? = null
+        fun getInstance() = instance ?: CoinRemoteDataSource().also { instance = it }
```
Comment (thonc-2135) [inline app/src/main/java/com/sunasterisk/coinqapp/data/source/remote/CoinRemoteDataSource.kt]: giãn cách giữa biến và hàm 1 dòng trống

File: app/src/main/java/com/sunasterisk/coinqapp/utils/getJsonFromURL.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    var inputStreamReader: InputStreamReader? = null
+    var buffer: BufferedReader? = null
+    try {
+        connection = url.openConnection() as HttpURLConnection
+        inputStreamReader = InputStreamReader(connection.inputStream)
+        buffer = BufferedReader(inputStreamReader)
+        buffer.forEachLine { builder.append(it) }
+    } catch (e: IOException) {
+        return builder.toString()
+    } finally {
```
Comment (thonc-2135) [inline app/src/main/java/com/sunasterisk/coinqapp/utils/getJsonFromURL.kt]: sao lỗi vẫn trả về builder à ?. Mà soa tên file lại đi đặt tên kiểu kia ?

File: app/src/main/java/com/sunasterisk/coinqapp/data/source/CoinDataSource.kt [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+import com.sunasterisk.coinqapp.data.model.Coin
+import com.sunasterisk.coinqapp.data.model.Exchange
+import com.sunasterisk.coinqapp.utils.OnLoadDataCallBack
+import java.util.*
+
+interface CoinDataSource {
+
+    interface Remote {
+        fun getCoins(
```
Comment (datdt-1935) [inline app/src/main/java/com/sunasterisk/coinqapp/data/source/CoinDataSource.kt]: Nhiều param quá thì nên tạo 1 class Request để gửi lên nhé

**Pull request merged**